File: mock_project/cpp_legacy/src/legacy_system.cpp
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <iostream>
#include <string>
#include <fstream>
#include <algorithm> // Add for std::sort
#include "../include/legacy_system.h"  // Include the header file

// Global variables
std::string g_errorMessage = "";
int g_debugMode = 0;
int g_itemCount = 0;
bool g_isInitialized = false;

// Global arrays
User* g_users[MAX_USERS];
Item g_items[MAX_ITEMS];
int g_userCount = 0;

// Function prototypes
void initSystem();
void cleanupSystem();
int addUser(int id, const std::string& name, const std::string& description, int age);
bool removeUser(int id);
void displayUserInfo(int id);
Item* findItem(int itemId);
User* getUserById(int id);
void setError(const std::string& message);
// The error is that `std::string` is used, but the necessary header `<string>` (which defines `std::string`) 
// has likely not been included in this .cpp file. This causes the compiler to report an error,
// such as 'namespace "std" has no member "string"', because it cannot find the definition for `std::string`.
void processData(const std::string& data, int (*processFn)(const std::string&, int));
bool saveToFile(const std::string& filename);
bool loadFromFile(const std::string& filename);
void sortItems();

// Initialize the system
void initSystem() {
    if (g_isInitialized) {
        return;
    }
    
    for (int i = 0; i < MAX_USERS; i++) {
        g_users[i] = nullptr;
    }
    
    for (int i = 0; i < MAX_ITEMS; i++) {
        g_items[i].id = -1;
    }
    
    std::srand(std::time(nullptr));
    g_isInitialized = true;
    std::cout << "System initialized" << std::endl;
}

// Clean up resources
void cleanupSystem() {
    if (!g_isInitialized) {
        return;
    }
    
    for (int i = 0; i < MAX_USERS; i++) {
        if (g_users[i] != nullptr) {
            delete g_users[i];
            g_users[i] = nullptr;
        }
    }
    
    g_isInitialized = false;
    std::cout << "System cleaned up" << std::endl;
}

// Add a new user to the system
int addUser(int id, const std::string& name, const std::string& description, int age) {
    if (g_userCount >= MAX_USERS) {
        setError("Maximum number of users reached");
        return -1;
    }
    
    for (int i = 0; i < g_userCount; i++) {
        if (g_users[i] != nullptr && g_users[i]->id == id) {
            setError("User ID already exists");
            return -1;
        }
    }
    
    User* newUser = new User();
    if (newUser == nullptr) {
        setError("Memory allocation failed");
        return -1;
    }
    
    newUser->id = id;
    newUser->name = name;
    newUser->description = description;
    newUser->age = age;
    newUser->isActive = true;
    
    g_users[g_userCount] = newUser;
    g_userCount++;
    
    return id;
}

// Remove a user from the system
bool removeUser(int id) {
    for (int i = 0; i < g_userCount; i++) {
        if (g_users[i] != nullptr && g_users[i]->id == id) {
            delete g_users[i];
            
            // Shift remaining users
            for (int j = i; j < g_userCount - 1; j++) {
                g_users[j] = g_users[j + 1];
            }
            
            g_users[g_userCount - 1] = nullptr;
            g_userCount--;
            
            return true;
        }
    }
    
    setError("User not found");
    return false;
}

// Display user information
void displayUserInfo(int id) {
    User* user = getUserById(id);
    if (user == nullptr) {
        std::cout << "User not found" << std::endl;
        return;
    }
    
    std::cout << "User ID: " << user->id << std::endl;
    std::cout << "Name: " << user->name << std::endl;
    std::cout << "Description: " << user->description << std::endl;
    std::cout << "Age: " << user->age << std::endl;
    std::cout << "Status: " << (user->isActive ? "Active" : "Inactive") << std::endl;
}

// Find an item by ID using std::find_if
Item* findItem(int itemId) {
    auto it = std::find_if(g_items, g_items + g_itemCount,
                         [itemId](const Item& item) { return item.id == itemId; });
    
    return (it != g_items + g_itemCount) ? &(*it) : nullptr;
}

// Get user by ID using std::find_if
User* getUserById(int id) {
    auto it = std::find_if(g_users, g_users + g_userCount,
                         [id](const User* user) { return user != nullptr && user->id == id; });
    
    return (it != g_users + g_userCount) ? *it : nullptr;
}

// Set error message
void setError(const std::string& message) {
    g_errorMessage = message;
    if (g_debugMode) {
        std::cerr << "ERROR: " << g_errorMessage << std::endl;
    }
}

// Get error message
std::string getError() {
    return g_errorMessage;
}

// Process data with callback function
void processData(const std::string& data, int (*processFn)(const std::string&, int)) {
    std::string buffer = data;
    
    int result = processFn(buffer, buffer.length());
    
    if (result != 0) {
        std::cout << "Data processing failed with code: " << result << std::endl;
    }
}

// Save data to file
bool saveToFile(const std::string& filename) {
    std::ofstream file(filename, std::ios::binary);
    if (!file) {
        setError("Failed to open file for writing");
        return false;
    }
    
    file.write(reinterpret_cast<const char*>(&g_userCount), sizeof(int));
    
    for (int i = 0; i < g_userCount; i++) {
        User* user = g_users[i];
        
        file.write(reinterpret_cast<const char*>(&user->id), sizeof(int));
        
        // Write name
        int nameLength = static_cast<int>(user->name.length());
        file.write(reinterpret_cast<const char*>(&nameLength), sizeof(int));
        file.write(user->name.c_str(), nameLength);
        
        // Write description
        int descriptionLength = static_cast<int>(user->description.length());
        file.write(reinterpret_cast<const char*>(&descriptionLength), sizeof(int));
        file.write(user->description.c_str(), descriptionLength);
        
        file.write(reinterpret_cast<const char*>(&user->age), sizeof(int));
        file.write(reinterpret_cast<const char*>(&user->isActive), sizeof(bool));
    }
    
    file.write(reinterpret_cast<const char*>(&g_itemCount), sizeof(int));
    
    for (int i = 0; i < g_itemCount; i++) {
        Item& item = g_items[i];
        
        file.write(reinterpret_cast<const char*>(&item.id), sizeof(int));
        
        // Write name
        int nameLength = static_cast<int>(item.name.length());
        file.write(reinterpret_cast<const char*>(&nameLength), sizeof(int));
        file.write(item.name.c_str(), nameLength);
        
        file.write(reinterpret_cast<const char*>(&item.price), sizeof(float));
        file.write(reinterpret_cast<const char*>(&item.quantity), sizeof(int));
        file.write(reinterpret_cast<const char*>(&item.owner_id), sizeof(int));
    }
    
    return true;
}

// Load data from file
bool loadFromFile(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        setError("Failed to open file for reading");
        return false;
    }
    
    // Clean up existing data
    cleanupSystem();
    initSystem();
    
    file.read(reinterpret_cast<char*>(&g_userCount), sizeof(int));
    
    for (int i = 0; i < g_userCount; i++) {
        User* user = new User();
        if (user == nullptr) {
            setError("Memory allocation failed during load");
            return false;
        }
        
        file.read(reinterpret_cast<char*>(&user->id), sizeof(int));
        
        // Read name
        int nameLength;
        file.read(reinterpret_cast<char*>(&nameLength), sizeof(int));
        char* nameBuf = new char[nameLength + 1];
        file.read(nameBuf, nameLength);
        nameBuf[nameLength] = '\0';
        user->name = nameBuf;
        delete[] nameBuf;
        
        // Read description
        int descriptionLength;
        file.read(reinterpret_cast<char*>(&descriptionLength), sizeof(int));
        char* descBuf = new char[descriptionLength + 1];
        file.read(descBuf, descriptionLength);
        descBuf[descriptionLength] = '\0';
        user->description = descBuf;
        delete[] descBuf;
        
        file.read(reinterpret_cast<char*>(&user->age), sizeof(int));
        file.read(reinterpret_cast<char*>(&user->isActive), sizeof(bool));
        
        g_users[i] = user;
    }
    
    file.read(reinterpret_cast<char*>(&g_itemCount), sizeof(int));
    
    for (int i = 0; i < g_itemCount; i++) {
        file.read(reinterpret_cast<char*>(&g_items[i].id), sizeof(int));
        
        // Read name
        int nameLength;
        file.read(reinterpret_cast<char*>(&nameLength), sizeof(int));
        char* nameBuf = new char[nameLength + 1];
        file.read(nameBuf, nameLength);
        nameBuf[nameLength] = '\0';
        g_items[i].name = nameBuf;
        delete[] nameBuf;
        
        file.read(reinterpret_cast<char*>(&g_items[i].price), sizeof(float));
        file.read(reinterpret_cast<char*>(&g_items[i].quantity), sizeof(int));
        file.read(reinterpret_cast<char*>(&g_items[i].owner_id), sizeof(int));
    }
    
    return true;
}

// Sort items by price using std::sort
void sortItems() {
    std::sort(g_items, g_items + g_itemCount, 
        [](const Item& a, const Item& b) {
            return a.price < b.price;
        });
}

// Example of a complex function with many parameters
void updateItemDetails(int itemId, const std::string& newName, float newPrice, int newQuantity, bool notifyOwner) {
    Item* item = findItem(itemId);
    if (item == nullptr) {
        setError("Item not found");
        return;
    }
    
    item->name = newName;
    item->price = newPrice;
    item->quantity = newQuantity;
    
    if (notifyOwner) {
        User* owner = getUserById(item->owner_id);
        if (owner != nullptr) {
            std::cout << "Notification sent to " << owner->name << " about item update" << std::endl;
        }
    }
}

// Main function demonstration
int main() {
    initSystem();
    
    addUser(1, "John Doe", "First test user", 35);
    addUser(2, "Jane Smith", "Second test user with management access", 28);
    
    // Initialize items using direct assignment
    g_items[g_itemCount].id = 1;
    g_items[g_itemCount].name = "Laptop";
    g_items[g_itemCount].price = 999.99;
    g_items[g_itemCount].quantity = 10;
    g_items[g_itemCount].owner_id = 1;
    g_itemCount++;
    
    g_items[g_itemCount].id = 2;
    g_items[g_itemCount].name = "Smartphone";
    g_items[g_itemCount].price = 499.99;
    g_items[g_itemCount].quantity = 20;
    g_items[g_itemCount].owner_id = 2;
    g_itemCount++;
    
    displayUserInfo(1);
    
    // Using std::string instead of C-style strings avoids memory leaks
    std::string tempDescription = "Temporary description";
    std::cout << "Temp: " << tempDescription << std::endl;
    
    // Safe buffer handling with std::string
    std::string safeBuffer = "This string can be of any length without buffer overflow risk";
    std::cout << "Safe buffer: " << safeBuffer << std::endl;
    
    // Modern type casting
    float f = 3.14;
    int i = static_cast<int>(f);
    std::cout << "Converted float to int: " << i << std::endl;
    
    sortItems();
    saveToFile("data.bin");
    
    cleanupSystem();
    return 0;
} 

File: TypescriptCode/legacy_system.ts
import * as fs from 'fs';
import * as path from 'path';

const MAX_USERS = 100;
const MAX_NAME_LENGTH = 50; // Will be used for validation, not strict char arrays
const MAX_ITEMS = 200;
// const BUFFER_SIZE = 1024; // Less relevant with dynamic strings

interface User {
    id: number;
    name: string;
    description: string;
    age: number;
    isActive: boolean;
}

interface Item {
    id: number;
    name: string;
    price: number;
    quantity: number;
    ownerId: number; // Changed from owner_id for camelCase convention
}

type ProcessFunction = (data: string, length: number) => number;

class LegacySystem {
    private errorMessage: string = "";
    private debugMode: boolean = false;
    private itemCount: number = 0;
    private isInitialized: boolean = false;

    private users: User[] = [];
    private items: Item[] = [];
    // userCount can be derived from users.length

    constructor(debugMode: boolean = false) {
        this.debugMode = debugMode;
    }

    public getUsers(): ReadonlyArray<User> {
        return this.users;
    }

    public getItems(): ReadonlyArray<Item> {
        return this.items;
    }

    public initSystem(): void {
        if (this.isInitialized) {
            return;
        }
        this.users = [];
        this.items = [];
        this.itemCount = 0;
        // srand(time(NULL)) equivalent for random seeding is not directly needed
        // unless specific predictable random numbers are required.
        this.isInitialized = true;
        console.log("System initialized");
    }

    public cleanupSystem(): void {
        if (!this.isInitialized) {
            return;
        }
        this.users = [];
        this.items = [];
        this.itemCount = 0;
        // No manual memory management (free) needed for users or descriptions
        this.isInitialized = false;
        console.log("System cleaned up");
    }

    private setError(message: string): void {
        this.errorMessage = message;
        if (this.debugMode) {
            console.error(`ERROR: ${this.errorMessage}`);
        }
    }

    public getError(): string {
        return this.errorMessage;
    }

    public addUser(id: number, name: string, description: string, age: number): number {
        if (this.users.length >= MAX_USERS) {
            this.setError("Maximum number of users reached");
            return -1;
        }

        if (this.users.some(user => user.id === id)) {
            this.setError("User ID already exists");
            return -1;
        }

        if (name.length >= MAX_NAME_LENGTH) {
            console.warn(`Warning: User name for ID ${id} exceeds MAX_NAME_LENGTH and will be truncated.`);
            name = name.substring(0, MAX_NAME_LENGTH -1);
        }

        const newUser: User = {
            id,
            name,
            description,
            age,
            isActive: true,
        };

        this.users.push(newUser);
        return id;
    }

    public removeUser(id: number): boolean {
        const userIndex = this.users.findIndex(user => user.id === id);
        if (userIndex === -1) {
            this.setError("User not found");
            return false;
        }

        this.users.splice(userIndex, 1);
        // No need to manually shift elements or nullify last element
        return true;
    }

    public displayUserInfo(id: number): void {
        const user = this.getUserById(id);
        if (!user) {
            console.log("User not found");
            return;
        }

        console.log(`User ID: ${user.id}`);
        console.log(`Name: ${user.name}`);
        console.log(`Description: ${user.description}`);
        console.log(`Age: ${user.age}`);
        console.log(`Status: ${user.isActive ? "Active" : "Inactive"}`);
    }

    public findItem(itemId: number): Item | null {
        const item = this.items.find(item => item.id === itemId);
        return item || null;
    }

    public getUserById(id: number): User | null {
        const user = this.users.find(user => user.id === id);
        return user || null;
    }
    
    public processData(data: string, processFn: ProcessFunction): void {
        // Buffer copy is not strictly necessary as strings are immutable
        // but to mimic the original's intent of working on a copy:
        let buffer = data; 
        const result = processFn(buffer, buffer.length);

        if (result !== 0) {
            console.log(`Data processing failed with code: ${result}`);
        } else {
            console.log(`Data processing succeeded.`);
        }
    }

    public saveToFile(filename: string): boolean {
        try {
            const dataToSave = {
                users: this.users,
                items: this.items,
                itemCount: this.itemCount // Or derive from items.length upon load
            };
            fs.writeFileSync(filename, JSON.stringify(dataToSave, null, 2), 'utf8');
            return true;
        } catch (error) {
            this.setError(`Failed to save file: ${error instanceof Error ? error.message : String(error)}`);
            return false;
        }
    }

    public loadFromFile(filename: string): boolean {
        try {
            if (!fs.existsSync(filename)) {
                this.setError("File not found for loading.");
                return false;
            }
            const fileContent = fs.readFileSync(filename, 'utf8');
            const dataLoaded = JSON.parse(fileContent);

            this.cleanupSystem(); // Clean existing data
            this.initSystem();   // Reinitialize

            this.users = dataLoaded.users || [];
            this.items = dataLoaded.items || [];
            this.itemCount = dataLoaded.items?.length || 0; // Recalculate item count

            return true;
        } catch (error) {
            this.setError(`Failed to load file: ${error instanceof Error ? error.message : String(error)}`);
            // Perform cleanup if loading fails mid-way to ensure consistent state
            this.cleanupSystem();
            this.initSystem();
            return false;
        }
    }

    public sortItems(): void {
        this.items.sort((a, b) => a.price - b.price);
    }
    
    public updateItemDetails(itemId: number, newName: string, newPrice: number, newQuantity: number, notifyOwner: boolean): void {
        const item = this.findItem(itemId);
        if (!item) {
            this.setError("Item not found");
            return;
        }

        if (newName.length >= MAX_NAME_LENGTH) {
            console.warn(`Warning: Item name for ID ${itemId} exceeds MAX_NAME_LENGTH and will be truncated.`);
            item.name = newName.substring(0, MAX_NAME_LENGTH -1);
        } else {
            item.name = newName;
        }
        item.price = newPrice;
        item.quantity = newQuantity;

        if (notifyOwner) {
            const owner = this.getUserById(item.ownerId);
            if (owner) {
                console.log(`Notification sent to ${owner.name} about item update for item ID ${item.id}`);
            } else {
                console.log(`Owner (ID: ${item.ownerId}) not found for item ID ${item.id}, notification not sent.`);
            }
        }
    }

    // Helper to add items directly for demonstration, similar to C++ main
    public addItem(id: number, name: string, price: number, quantity: number, ownerId: number): void {
        if (this.items.length >= MAX_ITEMS) {
            this.setError("Maximum number of items reached.");
            return;
        }
        if (this.items.some(item => item.id === id)) {
            this.setError(`Item with ID ${id} already exists.`);
            return;
        }
         if (name.length >= MAX_NAME_LENGTH) {
            console.warn(`Warning: Item name for ID ${id} exceeds MAX_NAME_LENGTH and will be truncated.`);
            name = name.substring(0, MAX_NAME_LENGTH -1);
        }
        this.items.push({ id, name, price, quantity, ownerId });
        this.itemCount = this.items.length;
    }
}

// Main function demonstration
function main(): void {
    const system = new LegacySystem(true); // Enable debug mode
    system.initSystem();

    const user1Id = system.addUser(1, "John Doe", "First test user", 35);
    if (user1Id === -1) console.error(system.getError());
    
    const user2Id = system.addUser(2, "Jane Smith", "Second test user with management access", 28);
    if (user2Id === -1) console.error(system.getError());

    system.addItem(1, "Laptop", 999.99, 10, 1);
    system.addItem(2, "Smartphone", 499.99, 20, 2);
    system.addItem(3, "Tablet", 299.00, 15, 1);


    if (user1Id !== -1) system.displayUserInfo(user1Id);
    if (user2Id !== -1) system.displayUserInfo(user2Id);
    
    console.log("\nItem List (Unsorted):");
    system.getItems().forEach(item => console.log(item));
    
    system.sortItems();
    console.log("\nItem List (Sorted by Price):");
    system.getItems().forEach(item => console.log(item));

    system.updateItemDetails(1, "Gaming Laptop", 1199.99, 8, true);
    const updatedItem = system.findItem(1);
    if(updatedItem) console.log("\nUpdated Item:", updatedItem);

    const dataFilePath = path.join(__dirname, 'data.json');
    console.log(`\nSaving data to ${dataFilePath}...`);
    if (system.saveToFile(dataFilePath)) {
        console.log("Data saved successfully.");
    } else {
        console.error("Failed to save data:", system.getError());
    }

    // Demonstrate loading
    console.log("\nCleaning system and loading data...");
    system.cleanupSystem(); // Clean current state
    system.initSystem();    // Re-init before load
    
    if (system.loadFromFile(dataFilePath)) {
        console.log("Data loaded successfully.");
        console.log("Users after load:", system.getUsers());
        console.log("Items after load:", system.getItems());
        const users = system.getUsers();
        if (users.length > 0) system.displayUserInfo(users[0].id);
    } else {
        console.error("Failed to load data:", system.getError());
    }

    // Example for processData
    const sampleData = "This is some sample data to process.";
    const myProcessFn: ProcessFunction = (d, l) => {
        console.log(`Processing data: "${d}" (length: ${l})`);
        // Example: return 0 for success, non-zero for failure
        return d.includes("error") ? 1 : 0; 
    };
    system.processData(sampleData, myProcessFn);
    system.processData("This data contains an error trigger.", myProcessFn);


    system.cleanupSystem();
    
    // Remove the data file after demo
    try {
        if(fs.existsSync(dataFilePath)) fs.unlinkSync(dataFilePath);
    } catch (e) {
        console.error("Could not delete test data file", e)
    }
}

// Run the main demonstration
main();

// To run this:
// 1. Save as legacy_system.ts in a directory (e.g., TypescriptCode)
// 2. Make sure you have Node.js and TypeScript installed.
// 3. Install types for node: `npm install --save-dev @types/node` (or `yarn add --dev @types/node`) in your project
// 4. Compile: `tsc TypescriptCode/legacy_system.ts`
// 5. Run: `node TypescriptCode/legacy_system.js` 